tests/generators/cpp/src/test.cpp

```diff
@@ -14,3 +17,337 @@
+    using openassetio_traitgen_test_all::traits::aNamespace::AllPropertiesTrait;
+
+    const openassetio::TraitsDataPtr traitsData =
+        openassetio::TraitsData::make({AllPropertiesTrait::kId});
+
+    traitsData->setTraitProperty(AllPropertiesTrait::kId, Fixture::kPropertyKey,
+                                 Fixture::kUnexpectedValue);
+
+    const Fixture fixture{AllPropertiesTrait{traitsData}};
+
+    WHEN("property is queried without a default") {
+      THEN("exception is thrown") {
+        std::string expectedMsg = "Invalid stored value type: should be '";
+        expectedMsg += Fixture::kPropertyType;
+        expectedMsg += "'.";
+
+        CHECK_THROWS_MATCHES(fixture.getProperty(), std::runtime_error,
+                             Catch::Matchers::Message(expectedMsg));
+      }
+    }
+
+    WHEN("property is queried with a default") {
+      const PropertyType value = fixture.getProperty(Fixture::kDefaultValue);
+
+      THEN("default is returned") { CHECK(value == Fixture::kDefaultValue); }
+    }
+  }
+}
+
+TEMPLATE_TEST_CASE("Property setters", "", openassetio::Bool, openassetio::Int, openassetio::Float,
+                   openassetio::Str) {
+  using openassetio_traitgen_test_all::traits::aNamespace::AllPropertiesTrait;
+
+  using PropertyType = TestType;  // Catch2-injected template param.
+  using Fixture = PropertyFixture<PropertyType>;
+
+  GIVEN("an AllPropertiesTrait view of a blank TraitsData") {
+    const openassetio::TraitsDataPtr traitsData = openassetio::TraitsData::make();
+    Fixture fixture{AllPropertiesTrait{traitsData}};
+
+    WHEN("property is set") {
+      fixture.setProperty(Fixture::kExpectedValue);
+
+      THEN("TraitsData contains expected value") {
+        openassetio::trait::property::Value value;
+        [[maybe_unused]] const bool hasProp =
+            traitsData->getTraitProperty(&value, AllPropertiesTrait::kId, Fixture::kPropertyKey);
+
+        const PropertyType actualValue = std::get<PropertyType>(value);
+
+        CHECK(actualValue == Fixture::kExpectedValue);
+      }
+
+      THEN("TraitsData has been imbued") { CHECK(traitsData->hasTrait(AllPropertiesTrait::kId)); }
+    }
+  }
+}
+
+SCENARIO("Moveable property setters") {
+  using openassetio_traitgen_test_all::traits::aNamespace::AllPropertiesTrait;
+
+  GIVEN("an AllPropertiesTrait view of a blank TraitsData") {
+    const openassetio::TraitsDataPtr traitsData = openassetio::TraitsData::make();
+    AllPropertiesTrait trait{traitsData};
+
+    WHEN("string property is set with a moveable string") {
+      std::string value = "some string";
+      trait.setStringProperty(std::move(value));
+
+      THEN("string has been moved") {
+        // TODO(DF): There is currently no way to check that the value
+        //  was moved "all the way", e.g. check `.data()` pointer is the
+        //  same, since `TraitsData::getTraitProperty` always copies.
+        CHECK(value.empty());  // NOLINT(bugprone-use-after-move)
```

Comment:
[Consider] I think you may be testing an environmental implementation detail here. Moved-from objects are left in an unspecified state, and don't necessarily clear themselves. Whilst I find it hard to believe any conforming std::string implementation wouldn't behave in this manner, I don't think it's a guarantee.

Furthermore, I struggle to consider what `setStringProperty` could do here to influence whether this test passes or fails. Is there a signature or implementation of `setStringProperty` that could effect the outcome of the move?

Reply:
Yeah it's not super safe, but works in GCC and is only used in tests.  Perhaps a `STATIC_REQUIRE` could be devised that would ensure moveability.

This test can only check the signature of `setStringProperty` has a moveable parameter.  Inside `setStringProperty` we may well not then move it onward to the `TraitsData`, which is a limitation of the test and what I allude to in the TODO comment.

I crafted a little [Compiler Explorer](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIMwAcpK4AMngMmAByPgBGmMQS0gAOqAqETgwe3r7%2BQSlpjgJhEdEscQlStpj2hQxCBEzEBFk%2BfoFVNRn1jQTFUbHxibYNTS057QojveH9ZYNSAJS2qF7EyOwc5gDM4cjeWADUJltuk8ThwMfYJhoAgtu7%2B5hHJ16OtIQAnlc392Y7DD2XkOxzcogUSiaPzuvzOXgcBw8giYM2IRwA7FY7gccQdJugQCAzhc8QRiAB9Y5Y%2B7ogAiVN%2BvwAbqg8OgDixUEzMAAlTBUCBoBiTUkEolki7mABspOICwx1NxiNkKIixBMmPxhM53IgZwWGvpW2phsZdxZbI5XMwADUxHqCGLiUZZfKNYrcUiGqj3VqQDrMA65abjbjGXSzfc7uECBzVRA3Zjfkr3cmlbi/c7gLKXrSjmYzEwYqYCwzsemcQG%2BQK/QGgwsDca0xWmBD4gQIGAwGcAHSYFhJAifBONj0V5sY%2BnlnGp6fpzMSl1nXP5wvF8xmMu3CuV6122gOsV1/WjidK1uQju9/uD4cNrcpiMwukcJa0TgAVl4fg4WlIqE4NxrGsPEVjWZ5th4UgCE0V8lgAaxALYpR7D8zHRLYti4LgAgATnRLgNClLh9E4SRv1g/9OF4BQQA0aDYKWOBYBgRAUFQAc6HichKDQTj6ASYAuALUgsCZPB1htPBMAAdwAeSSRhOCgmhaAIeJaIgGJKJicJGk%2BZTeF05hiE%2BOSYm0TAHEM0g%2BLYQQ5IYWgDN/XgsBYQxgHEVzRLwYgrMcblaJ8zBVCst4Nj/GNqkoj4YmIfSPCwSiJRYQylioAxgAUKTZIUpTuF4fhBBEMR2EqYr5CUNRKN0EiDCMFBgMsfQ8BiWjICWVBBwyYKaOqAKMhcBh3E8Vo9FCGZSnKPR8nSAQxj8Ei5tqPppvmDpBoEbpRjGnISLsLa6imNaBgqYYekWvRJh6U65gqJYFDA9YJDfT8KJ8gCOAOVQAilABaKVJAOYBkGQA5hJ7MwDggXBCBIfMsIWXgYNchtSEQj96PfDhyNINKsdIH8/y%2Bmi6IYtHSGYtiVgIJI3h4wUOKSLjiEiVgNl%2BgGgZBsGIbMKHeEwfAiHOAkSMq0rxAq2RFBUdQfLq0gZISpJ0tIjgvyJyivrkt56djVAqB%2Bv7AeB0Hwch6GIA8fj4kRrhkYprR0YQTAmCwBIEw1vGCfo4neFJ2xydRl2EJAKUth7SR0TMKUpQCAJCMCXCtlwjWtg%2BknqOduCNbMLPA5z0O8%2B5Yg0mcSQgA) test to demonstrate.

Reply:
> Perhaps a STATIC_REQUIRE could be devised that would ensure moveability.

Yeah, that would be idea, but it's never easy in c++. You want to check that the argument type, (probably using `std::is_move_constructable`. There are [ways](https://www.boost.org/doc/libs/1_40_0/libs/function_types/doc/html/boost_functiontypes/introduction.html) to get at the argument types of a method, but the question is whether it's worth it , likely not.

Reply:
Actually, it's easier than that - `clang-tidy` to the rescue!  It'll flag an error if you try a `std::move` would have no effect.

Reply:
Updated to rely on clang-tidy in 984b3b78